Wikipedia/assets/codemirror/codemirror-range-determination-bundle.js

```diff
@@ -0,0 +1,736 @@
+    }
+  }
+
+  // If selectionRange ends inside a markup item's closing markup, the returned range end will be moved to end of closing markup.
+  const selectionEndsInClosingMarkupOfItem = markupItems.find(item => selectionRange.endsInsideRange(item.closingMarkupRange(), true))
+  if (selectionEndsInClosingMarkupOfItem) {
+    newSelectionRange.endLocation = selectionEndsInClosingMarkupOfItem.closingMarkupRange().endLocation
+  } else {
+    // If selectionRange ends inside a markup item's opening markup, the returned range end will be moved to start of opening markup.
+    const selectionEndsInOpeningMarkupOfItem = markupItems.find(item => selectionRange.endsInsideRange(item.openingMarkupRange(), true))
+    if (selectionEndsInOpeningMarkupOfItem) {
+      newSelectionRange.endLocation = selectionEndsInOpeningMarkupOfItem.openingMarkupRange().startLocation
+    }
+  }
+  
+  return newSelectionRange
+}
+
+const canSplitMarkupAroundSelectionRange = (codeMirror) => splitMarkupAroundSelectionRange(codeMirror, true)
+
+const splitMarkupAroundSelectionRange = (codeMirror, evaluateOnly = false) => {
+  const selectionRange = getItemRangeFromSelection(codeMirror)
+
+  let markupItems = markupItemsForItemRangeLines(codeMirror, selectionRange)
+
+  const markupItemOpeningOrClosingMarkupIntersectsSelectionRange = (item) => item.openingMarkupRange().intersectsRange(selectionRange, false) || item.closingMarkupRange().intersectsRange(selectionRange, false)
+
+  const selectionIncludesAnyOpeningOrClosingMarkup = markupItems.find(markupItemOpeningOrClosingMarkupIntersectsSelectionRange) !== undefined
+
+  if (selectionIncludesAnyOpeningOrClosingMarkup) {
+    return false
+  }
+
+  const selectionIntersectsItemInnerRange = (item) => item.innerRange.intersectsRange(selectionRange, true)
+  markupItems = markupItems.filter(selectionIntersectsItemInnerRange)// === undefined
+  // Return if selection doesn't intersect with any markup items (selected word at end of line after last markup item etc).
+  if (markupItems.length === 0) {
+    return false
+  }
+  
+  if (evaluateOnly) {
+    return true
+  }
+
+  let markupRangesToAddBeforeSelection = []
+  let markupRangesToAddAfterSelection = []
+
+  markupItems.forEach(item => {
+    markupRangesToAddBeforeSelection.unshift(item.closingMarkupRange())
+    markupRangesToAddAfterSelection.push(item.openingMarkupRange())
+  })
+
+  let accumulatedLeftMarkup = getTextFromRanges(codeMirror, markupRangesToAddBeforeSelection)
+  let accumulatedRightMarkup = getTextFromRanges(codeMirror, markupRangesToAddAfterSelection)
+
+  // Work-around for cases which will result in empty markup items such as `<u></u>`.
+  // Injects text inside such markup items so they can be identified and pruned. 
+  // This also prevents a codemirror bug where it incorrectly tokenizes empty markup items.
+  const removeMeMarker = 'R!E@M#O$V%E'
```

Comment:
I'd be OK with the empty markup case if it means removing this workaround - not sure it's worth the added complexity and potential for leaving this string in the wikitext.

I'm noticing that if I have a bold, underlined selection:
<img width="229" alt="Screen Shot 2019-03-21 at 6 31 33 AM" src="https://user-images.githubusercontent.com/741327/54746672-29875c00-4ba3-11e9-9d48-9e7cc3dbd038.png">

Then tap clear markup, I end up with:
<img width="315" alt="Screen Shot 2019-03-21 at 6 31 44 AM" src="https://user-images.githubusercontent.com/741327/54746698-37d57800-4ba3-11e9-9dc7-e9da4a3c0049.png">

Same problem with bold & italics:
<img width="313" alt="Screen Shot 2019-03-21 at 6 36 05 AM" src="https://user-images.githubusercontent.com/741327/54746851-9f8bc300-4ba3-11e9-9ba1-b31de65378f2.png">

Reply:
Not as critical, but I'm noticing extra steps in the undo stack - If I have this selection:
<img width="250" alt="Screen Shot 2019-03-21 at 6 46 32 AM" src="https://user-images.githubusercontent.com/741327/54747372-1a091280-4ba5-11e9-80cb-f463433debd7.png">

Then clear formatting, it takes two undo steps to get back to the initial state.

Reply:
good catch will check!

Reply:
Fixed both! 

The first issue was a result of an underlying code mirror wiki text parsing bug where it incorrectly tokenized the closing markup for certain italic-nested-in-bold cases.

I was able to [detect and correct](https://github.com/wikimedia/wikipedia-ios/pull/2964/commits/fccba146ea9a44e59fe4f9c93987c5f0a298dc47) for these cases within the non-tag range determination code. Nice 'cause it means our `MarkupItem` will have correct ranges even when the underlying code mirror tokens don't.

Second issue with undo stack getting too many items is fixed here https://github.com/wikimedia/wikipedia-ios/pull/2964/commits/e93c51433f555a05fffa5c8609761f54d8af1271

Working on nixing the `R!E@M#O$V%E` marker now...

Reply:
Got rid of `R!E@M#O$V%E` :)